sound/soc/sof/probe.c

```diff
@@ -0,0 +1,290 @@
+	size_t size;
+	int ret;
+
+	*params = NULL;
+	*bytes = 0;
+
+	reply = kzalloc(SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+	if (!reply)
+		return -ENOMEM;
+	msg.rhdr.hdr.size = sizeof(msg);
+	msg.rhdr.hdr.cmd = SOF_IPC_GLB_PROBE | cmd;
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg.rhdr.hdr.cmd, &msg,
+			msg.rhdr.hdr.size, reply, SOF_IPC_MSG_MAX_SIZE);
+	if (ret < 0 || reply->rhdr.error < 0)
+		goto exit;
+
+	size = reply->rhdr.hdr.size -= sizeof(*reply);
+	if (size <= 0)
+		goto exit;
+
+	*params = kmemdup(&reply->dma[0], size, GFP_KERNEL);
+	if (!*params) {
+		ret = -ENOMEM;
+		goto exit;
+	}
+	*bytes = size;
+
+exit:
+	kfree(reply);
+	return ret;
+}
+
+/**
+ * sof_ipc_probe_get_dma - retrieve list of active injection dmas
+ * @sdev:	SOF sound device
+ * @dma:	Returned list of active dmas
+ * @num_dma:	Returned count of active dmas
+ *
+ * Host sends DMA_GET request to obtain list of injection dmas it
+ * can use to transfer data over with.
+ *
+ * Note that list contains only injection dmas as there is only one
+ * extractor (dma) and it is always assigned on probing init.
+ * DSP knows exactly where data from extraction probes is going to,
+ * which is not the case for injection where multiple streams
+ * could be engaged.
+ */
+int sof_ipc_probe_get_dma(struct snd_sof_dev *sdev,
+		struct sof_probe_dma **dma, size_t *num_dma)
+{
+	int ret;
+
+	ret = sof_ipc_probe_get(sdev, SOF_IPC_PROBE_DMA_GET,
+			(void **)dma, num_dma);
+	*num_dma /= sizeof(**dma);
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc_probe_get_dma);
+
+/**
+ * sof_ipc_probe_dma_attach - attach to specified dmas
+ * @sdev:	SOF sound device
+ * @dma:	List of streams (dmas) to attach to
+ * @num_dma:	Number of elements in @dma
+ *
+ * Contrary to extraction, injection streams are never assigned
+ * on init. Before attempting any data injection, host is responsible
+ * for specifying streams which will be later used to transfer data
+ * to connected probe points.
+ */
+int sof_ipc_probe_dma_attach(struct snd_sof_dev *sdev,
+		struct sof_probe_dma *dma, size_t num_dma)
+{
+	struct sof_ipc_probe_dma_set_params *msg;
+	struct sof_ipc_reply reply;
+	size_t bytes = sizeof(*dma) * num_dma;
+	int ret;
+
+	msg = kmalloc(sizeof(*msg) + bytes, GFP_KERNEL);
+	if (!msg)
+		return -ENOMEM;
+	msg->hdr.size = sizeof(*msg) + bytes;
+	msg->hdr.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_DMA_SET;
+	memcpy(&msg->dma[0], dma, bytes);
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg->hdr.cmd, msg, msg->hdr.size,
+			&reply, sizeof(reply));
+	kfree(msg);
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc_probe_dma_attach);
+
+/**
+ * sof_ipc_probe_dma_detach - detach from specified dmas
+ * @sdev:		SOF sound device
+ * @stream_tag:	List of stream tags to detach from
+ * @num_stream_tag:	Number of elements in @stream_tag
+ *
+ * Host sends DMA_DETACH request to free previously attached stream
+ * from being occupied for injection. Each detach operation should
+ * match equivalent DMA_SET. Detach only when all probes tied to
```

Comment:
why can't we use ATTACH/DETACH instead of SET/DETACH? It's just confusing for no good reason, and it makes your work less clear.